File: src/main/java/com/feiliks/blog/BlogController.java
package com.feiliks.blog;

import com.feiliks.common.AuthorizationException;
import com.feiliks.common.NotFoundException;
import com.feiliks.common.PasswordUtil;
import java.util.HashMap;
import java.util.Map;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;
import com.feiliks.common.dto.LoginDTO;
import com.feiliks.common.dto.PasswordDTO;
import com.feiliks.rms.entities.Tag;
import com.feiliks.common.entities.User;
import com.feiliks.rms.repositories.TagRepository;
import com.feiliks.common.repositories.UserRepository;
import java.util.Collection;
import java.util.Date;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

@Controller
@RequestMapping(value = "/blog")
public class BlogController {

    @Autowired
    private BlogRepository blogRepo;

    @Autowired
    private UserRepository userRepo;

    @Autowired
    private TagRepository tagRepo;

    @ExceptionHandler(AuthorizationException.class)
    protected String handleAuthorization(AuthorizationException ex) {
        return "redirect:/app/blog/logout";
    }

    @ExceptionHandler(NotFoundException.class)
    protected String handleNotFound(NotFoundException ex) {
        return "redirect:/app/blog/";
    }

    protected void checkAuth(HttpServletRequest req) {
        checkAuth(req, null);
    }

    protected void checkAuth(HttpServletRequest req, User user) {
        HttpSession session = req.getSession();
        if (session == null || session.getAttribute("username") == null) {
            throw new AuthorizationException();
        }
        if (user != null) {
            if (user.getUsername() == null
                    || !user.getUsername().equals(session.getAttribute("username"))) {
                throw new AuthorizationException();
            }
        }
    }

    protected User getUser(HttpServletRequest req) {
        HttpSession session = req.getSession();
        if (session == null) {
            return null;
        }
        Object username = session.getAttribute("username");
        if (username == null) {
            return null;
        }
        return userRepo.findByUsername((String) username);
    }

    protected void checkTags(Collection<Tag> original, Collection<Tag> updated) {
        Set<Tag> tagsToAdd = new HashSet<>(), tagsToRemove = new HashSet<>(),
                updatedFetched = new HashSet<>();
        for (Tag tag : updated) {
            String tname = tag.getName();
            if (tname == null || tname.trim().isEmpty()) {
                continue;
            }
            Tag old = tagRepo.findByName(tname);
            if (old == null) { // completely new tag
                old = tagRepo.save(tag);
                tagsToAdd.add(old);
                updatedFetched.add(old);
            } else {
                if (!original.contains(old)) { // existing new tag
                    tagsToAdd.add(old);
                }
                updatedFetched.add(old);
            }
        }
        for (Tag tag : original) {
            if (!updatedFetched.contains(tag)) {
                tagsToRemove.add(tag);
            }
        }
        original.addAll(tagsToAdd);
        original.removeAll(tagsToRemove);
    }

    @GetMapping
    public ModelAndView indexPage(HttpServletRequest req) {
        Map<String, Object> data = new HashMap<>();
        data.put("blogs", blogRepo.findPublished());
        data.put("user", getUser(req));
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("index", data);
    }

    @GetMapping("/admin")
    public ModelAndView adminPage(
            HttpServletRequest req,
            @RequestParam(name = "page", required = false) Integer page) {
        checkAuth(req);
        Map<String, Object> data = new HashMap<>();
        User user = getUser(req);
        List<Blog> blogs = new LinkedList<>();
        PageRequest pager = new PageRequest(page == null || page < 1 ? 0 : page - 1, 10);
        if (user != null) {
            Page<Blog> blogsPage = blogRepo.findByOwner(user, pager);
            for (Blog blog : blogsPage.getContent()) {
                blogs.add(blog);
            }
        }
        data.put("page_number", pager.getPageNumber());
        data.put("page_size", pager.getPageSize());
        data.put("blogs", blogs);
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("admin", data);
    }

    @GetMapping("/edit")
    public String createBlogPage(HttpServletRequest req) {
        checkAuth(req);
        User user = getUser(req);
        Blog entity = blogRepo.findBySlugAndOwner(null, user);
        if (entity == null) {
            entity = new Blog();
            entity.setSlug(null);
            entity.setTitle("[new]");
            entity.setPublished(false);
            entity.setOwner(user);
            Date now = new Date();
            entity.setCreated(now);
            entity.setModified(now);
            entity = blogRepo.save(entity);
        }
        return "redirect:/app/blog/edit/" + entity.getId();
    }

    @GetMapping("/edit/{id}")
    public ModelAndView editBlogPage(
            HttpServletRequest req,
            @PathVariable Long id) {
        Blog blog = blogRepo.findOne(id);
        if (blog == null) {
            throw new NotFoundException(Blog.class, id.toString());
        }
        checkAuth(req, blog.getOwner());
        Map<String, Object> data = new HashMap<>();
        data.put("blog", new BlogDTO(blog));
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("edit", data);
    }

    @PostMapping("/edit/{id}")
    public String updateBlog(
            HttpServletRequest req,
            @PathVariable Long id,
            @ModelAttribute BlogDTO blog) {
        Blog original = blogRepo.findOne(id);
        if (original == null) {
            throw new NotFoundException(Blog.class, id.toString());
        }
        checkAuth(req, original.getOwner());
        Blog entity = blog.toEntity();
        original.setSlug(entity.getSlug());
        original.setPublished(entity.isPublished());
        original.setTitle(entity.getTitle());
        original.setContent(entity.getContent());
        original.setModified(new Date());
        checkTags(original.getTags(), entity.getTags());
        blogRepo.save(original);
        return "redirect:/app/blog/edit/" + id;
    }

    @GetMapping("/view/{slug}")
    public ModelAndView viewBlogPage(HttpServletRequest req, @PathVariable String slug) {
        Blog blog = blogRepo.findBySlug(slug);
        if (blog == null) {
            throw new NotFoundException(Blog.class, slug);
        }
        Map<String, Object> data = new HashMap<>();
        data.put("blog", blog);
        data.put("user", getUser(req));
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("blog", data);
    }

    @GetMapping("/login")
    public ModelAndView loginPage(HttpServletRequest req) {
        HttpSession session = req.getSession();
        if (session != null && session.getAttribute("username") != null) {
            return new ModelAndView("redirect:/app/blog/admin");
        }
        Map<String, Object> data = new HashMap<>();
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("login", data);
    }

    @PostMapping("/login")
    public String login(
            HttpServletRequest req,
            @ModelAttribute LoginDTO login,
            RedirectAttributes redirectAtts) {
        User user = userRepo.findByUsername(login.getUsername());
        if (user != null && PasswordUtil.hash(user.getUsername(), login.getPassword()).equals(user.getPassword())) {
            HttpSession session = req.getSession(true);
            session.setAttribute("username", login.getUsername());
            return "redirect:/app/blog/admin";
        }
        redirectAtts.addFlashAttribute("error_message", "Failed to login");
        return logout(req);
    }

    @GetMapping("/logout")
    public String logout(HttpServletRequest req) {
        HttpSession session = req.getSession();
        if (session != null) {
            session.invalidate();
        }
        return "redirect:/app/blog/login";
    }

    @GetMapping("/chpwd")
    public ModelAndView chPwdPage(HttpServletRequest req) {
        checkAuth(req);
        Map<String, Object> data = new HashMap<>();
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("chpwd", data);
    }

    @PostMapping("/chpwd")
    public String chPwd(
            HttpServletRequest req,
            @ModelAttribute PasswordDTO pass,
            RedirectAttributes redirectAtts) {
        checkAuth(req);
        User user = getUser(req);
        if (PasswordUtil.hash(user.getUsername(), pass.getOriginal()).equals(user.getPassword())) {
            user.setPassword(PasswordUtil.hash(user.getUsername(), pass.getPassword()));
            userRepo.save(user);
            redirectAtts.addFlashAttribute("message", "Your password has been updated.");
        } else {
            redirectAtts.addFlashAttribute("error_message", "Original password is incorrect.");
        }
        return "redirect:/app/blog/chpwd";
    }

    @GetMapping("/adduser")
    public ModelAndView addUserPage(HttpServletRequest req) {
        Map<String, Object> data = new HashMap<>();
        data.put("contextPath", req.getContextPath() + "/app/blog");
        return new ModelAndView("adduser", data);
    }

    @PostMapping("/adduser")
    public String addUser(
            HttpServletRequest req,
            @ModelAttribute LoginDTO login,
            RedirectAttributes redirectAtts) {
        User user = new User();
        user.setUsername(login.getUsername());
        user.setPassword(PasswordUtil.hash(login.getUsername(), login.getPassword()));
        userRepo.save(user);
        redirectAtts.addFlashAttribute("message", "User " + user.getUsername() + " is created.");
        return "redirect:/app/blog/adduser";
    }

}

File: src/main/kotlin/com/feiliks/blog/BlogController.kt
package com.feiliks.blog

import com.feiliks.blog.dtos.BlogDto
import com.feiliks.blog.entities.BlogEntity
import com.feiliks.blog.entities.TagEntity
import com.feiliks.blog.entities.UserEntity
import com.feiliks.blog.repositories.BlogRepository
import com.feiliks.blog.repositories.TagRepository
import com.feiliks.blog.repositories.UserRepository
import com.feiliks.common.AuthorizationException
import com.feiliks.common.NotFoundException
import com.feiliks.common.PasswordUtil
import com.feiliks.common.dto.LoginDto
import com.feiliks.common.dto.PasswordDto
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.mvc.support.RedirectAttributes
import java.util.*
import javax.servlet.http.HttpServletRequest
import kotlin.collections.HashMap
import kotlin.collections.HashSet


@Controller
@RequestMapping(value = ["/"])
class BlogController {

    @Autowired
    private lateinit var blogRepo: BlogRepository

    @Autowired
    private lateinit var userRepo: UserRepository

    @Autowired
    private lateinit var tagRepo: TagRepository

    @ExceptionHandler(AuthorizationException::class)
    protected fun handleAuthorization(ex: AuthorizationException) : String {
        return "redirect:/logout"
    }

    @ExceptionHandler(NotFoundException::class)
    protected fun handleNotFound(ex: NotFoundException): String {
        return "redirect:/"
    }

    protected fun checkAuth(req: HttpServletRequest, user: UserEntity? = null) {
        val session = req.session
        if (session?.getAttribute("username") == null)
            throw AuthorizationException()
        if (user != null) {
            if (user.username != session.getAttribute("username"))
                throw AuthorizationException()
        }
    }

    protected fun getUser(req: HttpServletRequest): UserEntity? {
        val username = req.session?.getAttribute("username") ?: return null
        return userRepo.findByUsername(username as String)
    }

    protected fun checkTags(
            original: MutableSet<TagEntity>,
            updated: Set<TagEntity>) {
        val tagsToAdd = HashSet<TagEntity>()
        val tagsToRemove = HashSet<TagEntity>()
        val updatedFetched = HashSet<TagEntity>()
        for (tag in updated) {
            val tname = tag.name.trim()
            if (tname.isEmpty()) continue
            var old = tagRepo.findByName(tname)
            if (old == null) { // new tag
                old = tagRepo.save(tag)
                tagsToAdd += old
                updatedFetched += old
            } else { // existing tag but not in original
                if (old !in original)
                    tagsToAdd += old
                updatedFetched += old
            }
        }
        for (tag in original) {
            if (tag !in updatedFetched)
                tagsToRemove += tag
        }
        original += tagsToAdd
        original -= tagsToRemove
    }

    @GetMapping
    fun indexPage(req: HttpServletRequest): ModelAndView {
        val data = HashMap<String, Any?>()
        data["blogs"] = blogRepo.findPublished()
        data["user"] = getUser(req)
        data["contextPath"] = req.contextPath
        return ModelAndView("index", data)
    }

    @GetMapping("/admin")
    fun adminPage(
            req: HttpServletRequest,
            @RequestParam(name = "page", required = false) page: Int?
    ): ModelAndView {
        checkAuth(req)
        val data = HashMap<String, Any?>()
        val user = getUser(req)
        val blogs = LinkedList<BlogEntity>()
        val p = page?.minus(1) ?: 0
        val pager = PageRequest.of(if (p > 0) p else 0, 10)
        if (user != null) {
            val blogsPage = blogRepo.findByOwner(user, pager)
            blogs += blogsPage.content
        }
        data["page_number"] = pager.pageNumber
        data["page_size"] = pager.pageSize
        data["blogs"] = blogs
        data["contextPath"] = req.contextPath
        return ModelAndView("admin", data)
    }

    @GetMapping("/edit")
    fun createBlogPage(req: HttpServletRequest): String {
        checkAuth(req)
        val user = getUser(req)
        var entity = blogRepo.findBySlugAndOwner(null, user)
        if (entity == null) {
            entity = BlogEntity()
            entity.slug = null
            entity.title = "[new]"
            entity.published = false
            entity.owner = user
            val now = Date()
            entity.created = now
            entity.modified = now
            entity = blogRepo.save(entity)
        }
        return "redirect:/edit/${entity.id}"
    }

    @GetMapping("/edit/{id}")
    fun editBlogPage(
            req: HttpServletRequest,
            @PathVariable id: Long
    ): ModelAndView {
        val blog = blogRepo.findById(id).orElseThrow {
            throw NotFoundException(BlogEntity::class.java, id.toString())
        }
        checkAuth(req, blog.owner)
        val data = HashMap<String, Any?>()
        data["blog"] = BlogDto(blog)
        data["contextPath"] = req.contextPath
        return ModelAndView("edit", data)
    }

    @PostMapping("/edit/{id}")
    fun updateBlog(
            req: HttpServletRequest,
            @PathVariable id: Long,
            @ModelAttribute blog: BlogDto
    ): String {
        val original = blogRepo.findById(id).orElseThrow {
            throw NotFoundException(BlogEntity::class.java, id.toString())
        }
        checkAuth(req, original.owner)
        original.slug = blog.slug
        original.published = blog.published
        original.title = blog.title
        original.content = blog.content
        original.modified = Date()
        checkTags(original.tags as MutableSet<TagEntity>, blog.tagsAsSet)
        blogRepo.save(original)
        return "redirect:/edit/$id"
    }

    @GetMapping("/view/{slug}")
    fun viewBlogPage(
            req: HttpServletRequest,
            @PathVariable slug: String): ModelAndView {
        val blog = blogRepo.findBySlug(slug) ?:
            throw NotFoundException(BlogEntity::class.java, slug)
        val data = HashMap<String, Any?>()
        data["blog"] = blog
        data["user"] = getUser(req)
        data["contextPath"] = req.contextPath
        return ModelAndView("blog", data)
    }

    @GetMapping("/login")
    fun loginPage(req: HttpServletRequest): ModelAndView {
        if (req.session?.getAttribute("username") != null)
            return ModelAndView("redirect:/admin")
        val data = HashMap<String, Any?>()
        data["contextPath"] = req.contextPath
        return ModelAndView("login", data)
    }

    @PostMapping("/login")
    fun login(
            req: HttpServletRequest,
            @ModelAttribute login: LoginDto,
            redirectAtts: RedirectAttributes
    ): String {
        val user = userRepo.findByUsername(login.username)
        if (user != null && PasswordUtil.hash(login.username!!, login.password!!) == user.password) {
            req.getSession(true).setAttribute("username", login.username);
            return "redirect:/admin"
        }
        redirectAtts.addFlashAttribute("error_message", "Failed to login")
        return logout(req)
    }

    @GetMapping("/logout")
    fun logout(req: HttpServletRequest): String {
        req.session?.invalidate()
        return "redirect:/login"
    }

    @GetMapping("/chpwd")
    fun chPwdPage(req: HttpServletRequest): ModelAndView {
        checkAuth(req)
        val data = HashMap<String, Any?>()
        data["contextPath"] = req.contextPath
        return ModelAndView("chpwd", data)
    }

    @PostMapping("/chpwd")
    fun chPwd(
            req: HttpServletRequest,
            @ModelAttribute pass: PasswordDto,
            redirectAtts: RedirectAttributes
    ): String {
        checkAuth(req)
        val user = getUser(req)
        if (user != null && PasswordUtil.hash(user.username, pass.original!!) == user.password) {
            user.password = PasswordUtil.hash(user.username, pass.password!!)
            userRepo.save(user)
            redirectAtts.addFlashAttribute("message", "Your password has been updated.")
        } else {
            redirectAtts.addFlashAttribute("error_message", "Original password is incorrect.")
        }
        return "redirect:/chpwd"
    }

    @GetMapping("/adduser")
    fun addUserPage(req: HttpServletRequest): ModelAndView {
        val data = HashMap<String, Any?>()
        data["contextPath"] = req.contextPath
        return ModelAndView("adduser", data)
    }

    @PostMapping("/adduser")
    fun addUser(
            req: HttpServletRequest,
            @ModelAttribute login: LoginDto,
            redirectAtts: RedirectAttributes
    ): String {
        val user = UserEntity()
        user.username = login.username!!
        user.password = PasswordUtil.hash(login.username!!, login.password!!)
        userRepo.save(user)
        redirectAtts.addFlashAttribute("message", "User ${user.username} is created.")
        return "redirect:/adduser"
    }

}
